sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/PartitionBasedLoadBalancer.java

```diff
@@ -0,0 +1,333 @@
+    }
+
+    /**
+     * This is the main method responsible for load balancing. This method is expected to be invoked by the {@link
+     * EventProcessor} periodically. Every call to this method will result in this {@link EventProcessor} owning <b>at
+     * most one</b> new partition.
+     * <p>
+     * The load is considered balanced when no active EventProcessor owns 2 partitions more than any other active
+     * EventProcessor.Given that each invocation to this method results in ownership claim of at most one partition,
+     * this algorithm converges gradually towards a steady state.
+     * </p>
+     * When a new partition is claimed, this method is also responsible for starting a partition pump that creates an
+     * {@link EventHubAsyncConsumer} for processing events from that partition.
+     */
+    public void loadBalance() {
+        /*
+         * Retrieve current partition ownership details from the datastore.
+         */
+        final Mono<Map<String, PartitionOwnership>> partitionOwnershipMono = partitionManager
+            .listOwnership(eventHubName, consumerGroupName)
+            .timeout(Duration.ofSeconds(1)) // TODO: configurable by the user
+            .collectMap(PartitionOwnership::partitionId, Function.identity());
+
+        /*
+         * Retrieve the list of partition ids from the Event Hub.
+         */
+        final Mono<List<String>> partitionsMono = eventHubAsyncClient
+            .getPartitionIds()
+            .timeout(Duration.ofSeconds(1)) // TODO: configurable
+            .collectList();
+
+        Mono.zip(partitionOwnershipMono, partitionsMono)
+            .map(this::loadBalance)
+            // if there was an error, log warning and TODO: call user provided error handler
+            .doOnError(ex -> logger.warning("Load balancing for event processor failed - {}", ex.getMessage()))
+            .subscribe();
+    }
+
+    /*
+     * This method works with the given partition ownership details and Event Hub partitions to evaluate whether the
+     * current Event Processor should take on the responsibility of processing more partitions.
+     */
+    private Mono<Void> loadBalance(final Tuple2<Map<String, PartitionOwnership>, List<String>> tuple) {
+
+        Map<String, PartitionOwnership> partitionOwnershipMap = tuple.getT1();
+        List<String> partitionIds = tuple.getT2();
+
+        if (ImplUtils.isNullOrEmpty(partitionIds)) {
+            // ideally, should never happen.
+            return Mono.error(new IllegalStateException("There are no partitions in Event Hub " + this.eventHubName));
+        }
+
+        if (!isValid(partitionOwnershipMap)) {
+            // User data is corrupt.
+            return Mono.error(new IllegalStateException("Invalid partitionOwnership data from PartitionManager"));
+        }
+
+        /*
+         * Remove all partitions ownerships that have not be modified for a long time. This means that the previous
```

Comment:
"A long time" is ambiguous.   You may wish to consider offering context around how that is determined.  "Remove ownership that have not been modified within the configured period" or similar?

_(also, "ownership" is both singular and plural)_